[es output] Replace_Dots not working in array elements

## Bug Report

**Describe the bug**

Replace_Dots works to replace dots in fields with underscores. However there is an edge case where it does not seem to work, that is when the dotted field name is in an array as follow:

`'{"foo": [{".dotted": "bar"}]}'`

- Steps to reproduce the problem:

The method I use requires docker.
Please save the following file as `docker-compose.yaml` and run `docker-compose up`:

```yaml
version: "3.7"

services:
  fluent-bit:
    image: fluent/fluent-bit
    command:
      - /fluent-bit/bin/fluent-bit
      - -i
      - dummy
      - -p
      - 'Dummy={"top": [{".dotted": "value"}]}'
      - -o
      - es
      - -p
      - "Host=elasticsearch"
      - -p
      - "Replace_Dots=On"
    depends_on:
      - elasticsearch
  elasticsearch:
    image: elasticsearch:7.6.2
    ports:
      - "9200:9200"
    environment:
      - discovery.type=single-node
```

Notice the elasticsearch container emit:

```text
elasticsearch_1  | "Caused by: java.lang.IllegalArgumentException: object field starting or ending with a [.] makes object resolution ambiguous: [.dotted]",
```

However if the dummy input is configured as `'Dummy={"top": {".dotted": "value"}}'` (without the array), there is no such error, and `Replace_Dots` works as expected.

Furthermore if the dot is elsewhere than the beggining, ES does ingest the log, but the dot is not replaced.

e.g. `[{"d.otted": "value"}, {"foo": "bar"}]`

yields a document in ES as:

```json
      {
        "_index" : "fluent-bit",
        "_type" : "flb_type",
        "_id" : "u_LhW3EBY51yOSkZgGsJ",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2020-04-08T22:20:36.001Z",
          "top" : [
            {
              "d.otted" : "value"
            },
            {
              "foo" : "bar"
            }
          ]
        }
      }
```

**Expected behavior**
I would expect a document as such to show up in ES, with the fist char dot replaced:

```json
     {
        "_index" : "fluent-bit",
        "_type" : "flb_type",
        "_id" : "bdPdW3EBCJwJktcCD1kS",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2020-04-08T22:15:45.004Z",
          "top" : [
            {
              "_dotted" : "value"
            },
            {
              "foo" : "bar"
            }
          ]
        }
```

Similarly, if the dot is in the middle of the field, the following would be expected:

```json
      {
        "_index" : "fluent-bit",
        "_type" : "flb_type",
        "_id" : "u_LhW3EBY51yOSkZgGsJ",
        "_score" : 1.0,
        "_source" : {
          "@timestamp" : "2020-04-08T22:20:36.001Z",
          "top" : [
            {
              "d_otted" : "value"
            },
            {
              "foo" : "bar"
            }
          ]
        }
      }
```

PS you can see es documents by running:

```sh
curl "localhost:9200/_search?pretty" \
  -H 'Content-Type: application/json' \
  -d'{ "query": { "match_all": {} }}'
```

You can delete this index by running:
```sh
curl -X DELETE "localhost:9200/fluent-bit?pretty"
```

**Screenshots**
<!--- If applicable, add screenshots to help explain your problem. -->
![image](https://user-images.githubusercontent.com/24660299/78838292-da85b200-79c3-11ea-9a1b-1133539acce0.png)


**Your Environment**
As per docker-compose.yaml above.

**Additional context**
For reasons beyond my control, there are services that emit json logs, that have fields in arrays that begin with `.`, they are not being output to elasticsearch due to this bug.

Sidenote: I've made a PR for local pipeline setup with docker as per above via https://github.com/fluent/fluent-bit-docs/pull/288

--- a/plugins/out_es/es.c
+++ b/plugins/out_es/es.c
@@ -37,6 +37,10 @@
 
 struct flb_output_plugin out_es_plugin;
 
+inline int es_pack_array_content(msgpack_packer *tmp_pck,
+                                 msgpack_object array,
+                                 struct flb_elasticsearch *ctx);
+
 #ifdef FLB_HAVE_AWS
 static flb_sds_t add_aws_auth(struct flb_http_client *c,
                               struct flb_elasticsearch *ctx)
@@ -145,12 +149,51 @@ static inline int es_pack_map_content(msgpack_packer *tmp_pck,
             msgpack_pack_map(tmp_pck, v->via.map.size);
             es_pack_map_content(tmp_pck, *v, ctx);
         }
+        /*
+         * The value can be any data type, if it's an array we need to
+         * pass it to es_pack_array_content.
+         */
+        else if (v->type == MSGPACK_OBJECT_ARRAY)
+        {
+          msgpack_pack_array(tmp_pck, v->via.array.size);
+          es_pack_array_content(tmp_pck, *v, ctx);
+        }
         else {
             msgpack_pack_object(tmp_pck, *v);
         }
     }
-
     return 0;
+}
+
+/*
+  * Iterate through the array and sanitize elements.
+  * Mutual recursion with es_pack_map_content.
+  */
+inline int es_pack_array_content(msgpack_packer *tmp_pck,
+                                 msgpack_object array,
+                                 struct flb_elasticsearch *ctx)
+{
+  msgpack_object *e;
+
+  for (int i = 0; i < array.via.array.size; i++)
+  {
+    e = &array.via.array.ptr[i];
+    if (e->type == MSGPACK_OBJECT_MAP)
+    {
+      msgpack_pack_map(tmp_pck, e->via.map.size);
+      es_pack_map_content(tmp_pck, *e, ctx);
+    }
+    else if (e->type == MSGPACK_OBJECT_ARRAY)
+    {
+      msgpack_pack_array(tmp_pck, e->via.array.size);
+      es_pack_array_content(tmp_pck, *e, ctx);
+    }
+    else
+    {
+      msgpack_pack_object(tmp_pck, *e);
+    }
+  }
+  return 0;
 }
 
 /*

--- a/plugins/out_es/es.c
+++ b/plugins/out_es/es.c
@@ -153,8 +153,7 @@ static int es_pack_map_content(msgpack_packer *tmp_pck,
          * The value can be any data type, if it's an array we need to
          * pass it to es_pack_array_content.
          */
-        else if (v->type == MSGPACK_OBJECT_ARRAY)
-        {
+        else if (v->type == MSGPACK_OBJECT_ARRAY) {
           msgpack_pack_array(tmp_pck, v->via.array.size);
           es_pack_array_content(tmp_pck, *v, ctx);
         }
@@ -173,27 +172,26 @@ static int es_pack_array_content(msgpack_packer *tmp_pck,
                                  msgpack_object array,
                                  struct flb_elasticsearch *ctx)
 {
-  msgpack_object *e;
+    msgpack_object *e;
 
-  for (int i = 0; i < array.via.array.size; i++)
-  {
-    e = &array.via.array.ptr[i];
-    if (e->type == MSGPACK_OBJECT_MAP)
-    {
-      msgpack_pack_map(tmp_pck, e->via.map.size);
-      es_pack_map_content(tmp_pck, *e, ctx);
+    for (int i = 0; i < array.via.array.size; i++) {
+        e = &array.via.array.ptr[i];
+        if (e->type == MSGPACK_OBJECT_MAP)
+        {
+            msgpack_pack_map(tmp_pck, e->via.map.size);
+            es_pack_map_content(tmp_pck, *e, ctx);
+        }
+        else if (e->type == MSGPACK_OBJECT_ARRAY)
+        {
+            msgpack_pack_array(tmp_pck, e->via.array.size);
+            es_pack_array_content(tmp_pck, *e, ctx);
+        }
+        else
+        {
+            msgpack_pack_object(tmp_pck, *e);
+        }
     }
-    else if (e->type == MSGPACK_OBJECT_ARRAY)
-    {
-      msgpack_pack_array(tmp_pck, e->via.array.size);
-      es_pack_array_content(tmp_pck, *e, ctx);
-    }
-    else
-    {
-      msgpack_pack_object(tmp_pck, *e);
-    }
-  }
-  return 0;
+    return 0;
 }
 
 /*

--- a/plugins/out_es/es.c
+++ b/plugins/out_es/es.c
@@ -37,7 +37,7 @@
 
 struct flb_output_plugin out_es_plugin;
 
-inline int es_pack_array_content(msgpack_packer *tmp_pck,
+static int es_pack_array_content(msgpack_packer *tmp_pck,
                                  msgpack_object array,
                                  struct flb_elasticsearch *ctx);
 
@@ -71,9 +71,9 @@ static flb_sds_t add_aws_auth(struct flb_http_client *c,
 }
 #endif /* FLB_HAVE_AWS */
 
-static inline int es_pack_map_content(msgpack_packer *tmp_pck,
-                                      msgpack_object map,
-                                      struct flb_elasticsearch *ctx)
+static int es_pack_map_content(msgpack_packer *tmp_pck,
+                               msgpack_object map,
+                               struct flb_elasticsearch *ctx)
 {
     int i;
     char *ptr_key = NULL;
@@ -169,7 +169,7 @@ static inline int es_pack_map_content(msgpack_packer *tmp_pck,
   * Iterate through the array and sanitize elements.
   * Mutual recursion with es_pack_map_content.
   */
-inline int es_pack_array_content(msgpack_packer *tmp_pck,
+static int es_pack_array_content(msgpack_packer *tmp_pck,
                                  msgpack_object array,
                                  struct flb_elasticsearch *ctx)
 {

